### DISCUSSION TITLE
error : A 'DarcyPressure' is not a registered object.

### DISCUSSION THREAD

**Discussion by ziyiwang141319** [idaholab/moose#26513] [Q&A Getting Started]
```
*** ERROR ***
A 'DarcyPressure' is not a registered object.

If you are trying to find this object in a dynamically loaded library, make sure that
the library can be found either in your "Problem/library_path" parameter or in the
MOOSE_LIBRARY_PATH environment variable.
```
I tried to create objects as problem and kernel, but always got this "not a registered object". So I copied DarcyPressure example in the tutorial, still got the same error. I created the head file(.h) and source file(.cpp) for DarcyPressure, and only changed the line`registerMooseObject("capybaraApp", DarcyPressure);` where capybara is my app name. I am using a mac system and I did make process before running inputfile.  I put the head file in include folder, and put source file in folder under src/kernels/.
Could you please help me with this?

**Comment by GiudGiud** [comment]
Hello

How is capybaraapp spelled in the base/ ... app.C file?

Any mismatch, caps included, between the registering and the name of the app will get you this error

Guillaume

**Reply by ziyiwang141319** [reply]
I did not put the first letter in capital when i created it. Here is the .cfile. 
```
#include "capybaraApp.h"
#include "Moose.h"
#include "AppFactory.h"
#include "ModulesApp.h"
#include "MooseSyntax.h"

InputParameters
capybaraApp::validParams()
{
  InputParameters params = MooseApp::validParams();
  params.set<bool>("use_legacy_material_output") = false;
  return params;
}

capybaraApp::capybaraApp(InputParameters parameters) : MooseApp(parameters)
{
  capybaraApp::registerAll(_factory, _action_factory, _syntax);
}

capybaraApp::~capybaraApp() {}

void 
capybaraApp::registerAll(Factory & f, ActionFactory & af, Syntax & s)
{
  ModulesApp::registerAllObjects<capybaraApp>(f, af, s);
  Registry::registerObjectsTo(f, {"capybaraApp"});
  Registry::registerActionsTo(af, {"capybaraApp"});

  /* register custom execute flags, action syntax, etc. here */
}

void
capybaraApp::registerApps()
{
  registerApp(capybaraApp);
}

/***************************************************************************************************
 *********************** Dynamic Library Entry Points - DO NOT MODIFY ******************************
 **************************************************************************************************/
extern "C" void
capybaraApp__registerAll(Factory & f, ActionFactory & af, Syntax & s)
{
  capybaraApp::registerAll(f, af, s);
}
extern "C" void
capybaraApp__registerApps()
{
  capybaraApp::registerApps();
}

```

Here is what I copied from tutorial as kernels
```
#include "DarcyPressure.h"

registerMooseObject("capybaraApp", DarcyPressure);

InputParameters
DarcyPressure::validParams()
{
  InputParameters params = ADKernelGrad::validParams();
  params.addClassDescription("Compute the diffusion term for Darcy pressure ($p$) equation: "
                             "$-\\nabla \\cdot \\frac{\\mathbf{K}}{\\mu} \\nabla p = 0$");

  return params;
}

DarcyPressure::DarcyPressure(const InputParameters & parameters)
  : ADKernelGrad(parameters),

    // Set the coefficients for the pressure kernel
    _permeability(0.8451e-09),
    _viscosity(7.98e-04)
{
}

ADRealVectorValue
DarcyPressure::precomputeQpResidual()
{
  return (_permeability / _viscosity) * _grad_u[_qp];
}

```

and header file I directly copied with no change 
```
#pragma once

// Including the "ADKernel" base class here so we can extend it
#include "ADKernelGrad.h"

/**
 * Computes the residual contribution: K / mu * grad_test * grad_u.
 */
class DarcyPressure : public ADKernelGrad
{
public:
  static InputParameters validParams();

  DarcyPressure(const InputParameters & parameters);

protected:
  /// ADKernel objects must override precomputeQpResidual
  virtual ADRealVectorValue precomputeQpResidual() override;

  /// The variables which hold the value for K and mu
  const Real _permeability;
  const Real _viscosity;
};
```

**Reply by GiudGiud** [reply]
ok so the name of the app is correct.
Where did you place this object source file? in your_app/src/kernels?
did you re-compile since you added the object? Did the compilation complete with no error message?

**Reply by YXY101676** [reply]
Hey 
GiudGiud
   I ran into the same problem, writing my own header and source files for material types and placing them within include and Src in my app.here is the headfile:
![1](https://github.com/idaholab/moose/assets/154942901/6cb71262-e9a5-45a6-af9a-1239d0c67661)
and my C.file also has  registerMooseObject("recrystallizationApp", NucleationProbability);
![image](https://github.com/idaholab/moose/assets/154942901/620c20ef-bdf0-41d8-8133-77112f8673af)
And the compilation complete with no error message,but  the same problem occurs 
![image](https://github.com/idaholab/moose/assets/154942901/20bba9c7-7d92-4413-81b0-7af6338992ac)

**Reply by GiudGiud** [reply]
Can you register this object to MooseApp for now?

**Reply by YXY101676** [reply]
Still can't， 
Is there a problem with the file format

**Reply by GiudGiud** [reply]
Your file must not be getting compiled. Are you sure you are using a newly compiled executable? And that the file is inside that app s folder?

Alternatively

git commit the file 
Then git clean -xfd the repository and rebuild after

**Reply by YXY101676** [reply]
It was in the folder：
![image](https://github.com/idaholab/moose/assets/154942901/21b354d8-f239-4521-97b8-3aa7419a62c7)
i donot know how to git commit the file

**Reply by YXY101676** [reply]
![image](https://github.com/idaholab/moose/assets/154942901/84d7d284-a82a-4dd3-9fd0-c711ffbe6f66)
I was just reinstalling moose，So I don't think the above command is going to work

**Reply by GiudGiud** [reply]
git add src
Git add include
Git commit

**Reply by YXY101676** [reply]
![image](https://github.com/idaholab/moose/assets/154942901/8deb3db3-88c2-4815-bb87-051549992d09)

**Reply by GiudGiud** [reply]
git not Git

Auto correct adds the capitalization

**Reply by YXY101676** [reply]
![image](https://github.com/idaholab/moose/assets/154942901/4de8d816-c5d6-40e8-9463-54e52bfd1764)
Then how to do it

**Reply by GiudGiud** [reply]
That s your default text editor

Probably vi or emacs

**Reply by YXY101676** [reply]
I don't know what to do next. I've never uploaded anything to github

**Reply by GiudGiud** [reply]
This does not upload it just saves your work 

There are good help pages for using git

**Reply by YXY101676** [reply]
ok，Thank you very much for your help
